### PR TITLE
Images by tag

### LINKED ISSUE #9: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [x] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by BinKuangHsiehApp**
Closes #9 
Hi, I'm actually waiting on the previous pull request but thought I can open this pull request so you can give me some feed back in the mean time. Relevant stuff should only be the last commit 4b59464

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -16,6 +16,7 @@ def create
 
   def index
     @images = Image.includes(:tags).order(created_at: :desc)
+    @images = @images.tagged_with(params[:tag]) if params[:tag]
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: What if the `params[:tag]` is invalid? And what if `@images.tagged_with(params[:tag])` returns empty?

File: app/views/images/index.html.erb [hunk 2]
```diff
@@ -8,7 +8,12 @@
               <%= image_tag image.image_link , class: 'image index_img' %>
             <% end %>
           </div>
-          <p class="card-text"><small class="text-muted">tags: <%=image.tags.join(',') %></small></p>
+          <p class="card-text">
+            <small class="text-muted">tags: </small>
+            <% image.tags.split(',') do |each_tag| %>
```
Comment (shiboying) [inline app/views/images/index.html.erb]: Minor, but name it `tag` can already imply it is a individual tag.

File: app/views/images/index.html.erb [hunk 3]
```diff
@@ -8,7 +8,12 @@
...
+              <small class="text-muted"><%= link_to each_tag.name, images_path(:params => {:tag => each_tag.name})%></small>
```
Comment (shiboying) [inline app/views/images/index.html.erb]: I would recommend you change `images_path(:params => {:tag => each_tag.name})` to `images_path(params: { tag => each_tag.name })`. And usually we put space after and before the `{}`.

File: app/views/images/show.html.erb [hunk 4]
```diff
@@ -1,3 +1,18 @@
...
+<div class='card'>
+  <div class="card-block ">
+    <div class='img_display card-img-top img-fluid text-center'>
+      <%= image_tag @current_image.image_link , class: 'image index_img' %>
+    </div>
+      <div class="card-block">
+      <p class="card-text">
+        <small class="text-muted">tags: </small>
+          <% @current_image.tags.split(',') do |each_tag| %>
+          <small class="text-muted tags"><%= link_to each_tag.name, images_path(:params => {:tag => each_tag.name})%></small>
```
Comment (shiboying) [inline app/views/images/show.html.erb]: Same here

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -82,6 +84,15 @@ def test_should_delete_image__should_not_delete_again
...
 
+  def test_should_filter_by_tag_when_clicked
+    blue_image = create_image(image_link: 'http://blue.png', tag_list: 'blue')
+    create_image(image_link: 'http://red.png', tag_list: 'red')
+    get images_path(tag: 'blue')
+    assert_select '.image', count: 1 do |elements|
+      assert_equal blue_image.image_link, elements[0].attribute('src').value
+    end
+  end
+
```
Comment (shiboying) [inline test/controllers/images_controller_test.rb]: You need to test the two other cases I mentioned

File: app/controllers/images_controller.rb (hunk 1 above)
Comment (BinKuangHsiehApp) [inline app/controllers/images_controller.rb]: I was thinking of using a Flash with redirect to the unfiltered index.

Comment (shiboying) [inline app/controllers/images_controller.rb]: That would be good

**Pull request merged**